diff --git a/node.gyp b/node.gyp
index 67909bdeb8..f1d9d15e5b 100644
--- a/node.gyp
+++ b/node.gyp
@@ -466,20 +466,8 @@
     },
 
     'conditions': [
-      # Pointer authentication for ARM64.
       ['target_arch=="arm64"', {
-          'target_conditions': [
-              ['_toolset=="host"', {
-                  'conditions': [
-                      ['host_arch=="arm64"', {
-                          'cflags': ['-mbranch-protection=standard'],
-                      }],
-                  ],
-              }],
-              ['_toolset=="target"', {
-                  'cflags': ['-mbranch-protection=standard'],
-              }],
-          ],
+        'cflags': ['-mbranch-protection=standard'],  # Pointer authentication.
       }],
       ['OS in "aix os400"', {
         'ldflags': [
