diff --git a/vcbuild.bat b/vcbuild.bat
index 9f324c7e5b8..761f4d0cc78 100644
--- a/vcbuild.bat
+++ b/vcbuild.bat
@@ -78,7 +78,7 @@ set exit_code=0
 :next-arg
 if "%1"=="" goto args-done
 if /i "%1"=="debug"         set config=Debug&goto arg-ok
-if /i "%1"=="release"       set config=Release&set ltcg=1&set cctest=1&goto arg-ok
+if /i "%1"=="release"       set config=Release&set cctest=1&goto arg-ok
 if /i "%1"=="clean"         set target=Clean&goto arg-ok
 if /i "%1"=="testclean"     set target=TestClean&goto arg-ok
 if /i "%1"=="ia32"          set target_arch=x86&goto arg-ok
@@ -174,7 +174,7 @@ if defined build_release (
   set i18n_arg=full-icu
   set projgen=1
   set cctest=1
-  set ltcg=1
+  set ltcg=0
 )
 
 if defined msi     set stage_package=1
